### PR TITLE
Use tikv-jemallocator in rustc/rustdoc in addition to jemalloc-sys when enabled.

### PR DESCRIPTION
**Pull request by guswynn**
In https://github.com/rust-lang/rust/pull/81782 it was mentioned that one reason rustc may benefit from minimalloc is it doesn't use the `sdallocx` api from jemalloc.

Currently, on unix, rust uses jemalloc by importing its symbols to use them with the default, System (libc) global allocator. 
This PR switches its global alloc to `tikv-jemallocator`, which correctly uses sized deallocation (https://docs.rs/tikv-jemallocator/0.4.1/src/tikv_jemallocator/lib.rs.html#121-126). `tikv-jemallocator`, as far as I can tell, is a more up-to-date set of bindings to jemalloc than `jemallocator`

The perf results of this pr are in large part due to the version upgrade of jemalloc, but sized deallocation has a non-trivial improvement, particularly to rustdoc.

This pr also includes changes to bootstrap to correctly pass the jemalloc feature through to the rustdoc build

### CODE REVIEW

**Comment by guswynn** [comment]
ci failure was just a tidy mistake (I ALWAYS forget to run that locally...)

**Comment by jyn514** [comment]
> Do we have perf tests for error paths? My hunch is that, while the clean path of rustc is heavily arena-driven, the error paths are more alloc-heavy

I don't think so, historically we haven't worried much about compile time for errors (see also https://rust-lang.zulipchat.com/#narrow/stream/242791-t-infra/topic/perfbot/near/229630033). I'd expect rustdoc to show any changes though, it does a *lot* of allocations.

**Comment by guswynn** [comment]
@jyn514 rustdoc doesn't appear to depend on this crate? In that case it may be using the libc allocator in all cases?

**Comment by jyn514** [comment]
Oh huh, I think you're right. Do you mind doing that here? You don't have to add jemalloc to cargo.toml, just load it from the sysroot (see the comment at the top of librustdoc/lib.rs).

**Comment by guswynn** [comment]
(I also testing with "jemalloc = true" in config.toml locally as even tho im on osx, its ci that turns that on, as well as used the more-correct jemalloc feature in rustc_driver)

File: src/librustdoc/lib.rs [hunk 1]
```diff
@@ -112,7 +118,45 @@ mod theme;
...
+        {
+            extern "C" {
+                fn _rjem_je_zone_register();
+            }
+
+            #[used]
+            static _F7: unsafe extern "C" fn() = _rjem_je_zone_register;
+        }
+    }
+
```
Comment (guswynn) [inline src/librustdoc/lib.rs:159]: not sure why this was originally in `rustc` and not `rustc_driver`, maybe it only works if its in the leaf binary crate? @alexcrichton is the origin of this I believe?

Comment (alexcrichton) [reply src/librustdoc/lib.rs:159]: The movement happened in https://github.com/rust-lang/rust/pull/56986 in response to https://github.com/rust-lang/rust/issues/56980. I believe it boils down to, at the time, "the RLS didn't do well with jemalloc". I don't believe the reason for "not doing well" was ever investigated, however.

Comment (guswynn) [reply src/librustdoc/lib.rs:159]: That encouraging to hear that it should correctly overwrite the symbols (and we can move to `src/tools/rustdoc/src/main.rs` if we need to for RLS (though I believe most people use rust-analyzer now)

**Comment by guswynn** [comment]
Does bor's need to re-queue with the new commits for rust-timer to catch up?

**Comment by jyn514** [comment]
> Does bors need to re-queue with the new commits for rust-timer to catch up?

Yes, otherwise rust-timer will use the build artifacts from the old commit.

@bors try @rust-timer queue

**Comment by LifeIsStrange** [comment]
It would probably be better to use this fork which is more up to date with Jemalloc:
https://github.com/tikv/jemallocator

**Comment by guswynn** [comment]
@LifeIsStrange I'll take a look!

**Comment by guswynn** [comment]
Definitely some improvements on docs, but not much else, I wonder if I did something wrong

**Comment by joshtriplett** [comment]
I don't think you did anything wrong.

It might be worth re-testing with tikv-jemallocator and tikv-jemalloc-sys, just to see how much of an improvement that provides. That'll give us a good baseline for comparison with mimalloc.

**Comment by guswynn** [comment]
I think only contributors can do bors and rust-timer?

File: src/librustdoc/lib.rs (hunk 1 above)
Comment (Xanewok) [reply src/librustdoc/lib.rs:163]: > I don't believe the reason for "not doing well" was ever investigated, however.

I don't think it was and IIRC the main problem was memory characteristics when running multiple `rustc` in-process in the RLS, which is not done anymore by default (we spawn child processes instead).

Feel free to disregard RLS when making changes here.

**Comment by Xanewok** [comment]
@bors try @rust-timer queue

(I'm not sure if I can use `delegate` for try/perf only)

**Comment by guswynn** [comment]
Those....look pretty good!

@LifeIsStrange what are the material differences between tikv-jemalloc* and upstream that would cause it to perform so much better? just the upgrade to jemalloc 5.2?

**Comment by joshtriplett** [comment]
Very impressive results! Not just the improvements to compile time, but massive improvements in memory usage.

**Comment by guswynn** [comment]
@joshtriplett oh i didnt even look at that!

**Comment by guswynn** [comment]
@joshtriplett appears to be mostly the upgrade to 5.2, but quite nice!

**Comment by joshtriplett** [comment]
Given the substantial improvement here, is there any reason we *shouldn't* merge this?

We can use this improvement as the new baseline for any further optimizations.

**Comment by Mark-Simulacrum** [comment]
It would be good to update the PR description.

Aside from that, I think one open question for me is whether the tikv jemalloc is the right one for us to use; do they have any patches atop baseline? Is it just more up to date?

**Comment by guswynn** [comment]
@Mark-Simulacrum from my cursory glance, patches are mostly a rename, simplification of CI, and the update the jemalloc version? @LifeIsStrange may know more

**Comment by sfackler** [comment]
It looks like this is the only change the jemalloc sources: https://github.com/tikv/jemalloc/commit/897abac9e71ece411ef6d326402bb6b2e1cee5f0

**Comment by LifeIsStrange** [comment]
Wow the max-rss wins are huge!
Given those wins and that jemalloc 5.3 should release ~soon I don't see a point in migrating to mimalloc for now however it should be ideally regularly  re-benchmarked as mimalloc might eventually take the performance crowd in future updates. Also it might shine in other benchmarcks (e.g for multithreading, etc) 

So looking at the commits of the tikv forks, most changes seems quite minor and the performance increase should be a direct result of the update of jemalloc, not of its wrappers. 
There was at least one noteworthy change though:
> Updated jemalloc to fix deadlock during initialization

Maybe that @BusyJay can tell us more about his work.

**Comment by BusyJay** [comment]
TL;DR: the [patch](https://github.com/tikv/jemalloc/commit/897abac9e71ece411ef6d326402bb6b2e1cee5f0) gets around the deadlock issue by registering a global hook lazily. It neither breaks original behavior nor fixes the issue completely.

The deadlock happens when calling `atexit` too many times before initializing jemalloc. Registering too many hooks using `atexit` will lead to allocation eventually. If jemalloc is not initialized, it will initialize itself. Before the patch, it will always calls `atexit` to register hook for stopping logging perf stats, hence deadlock in `atexit`. After this patch, the hook will be registered only when logging has been started. So it limits the deadlock conditions to register too many hooks using `atexit` and set `perf_log` to true.

**Comment by guswynn** [comment]
Does this pr need any other changes before its mergeable?

**Comment by jq-rs** [comment]
To me, this does not look that different compared to earlier #82389 measurements. Or do I miss something? And unfortunately, walltime is not improving in bootstrap.

**Comment by jyn514** [comment]
> To me, this does not look that different compared to earlier #82389 measurements.

The measurements in #82389 are also extremely good :) I wonder why you closed the PR?

> And unfortunately, walltime is not improving in bootstrap.

That's ok, we can't expect this to be better in *every* metric :laughing:

**Comment by jq-rs** [comment]
> > To me, this does not look that different compared to earlier #82389 measurements.
> 
> The measurements in #82389 are also extremely good :) I wonder why you closed the PR?

Well, I closed it as it was for reference only and walltime did not show improvement compared to `mimalloc` run :) As far as I can understand, here we have tested also the sized deallocation which appears to roughly match these results.

> > And unfortunately, walltime is not improving in bootstrap.
> 
> That's ok, we can't expect this to be better in _every_ metric laughing

Yes, this is indeed an improvement to earlier `jemalloc` version anyway.

**Comment by jq-rs** [comment]
It is a bit difficult to compare as the baseline is different but here is a [walltime comparison](https://perf.rust-lang.org/compare.html?start=029e25f8d5fcc8d7794c739999f25890290f416b&end=33fa9f15639c420e2e76b65192553337d67a7e0a&stat=wall-time) between these two runs. Here I'd say this run looks better indeed for walltime in general, although there is some variation.

**Comment by guswynn** [comment]
@jq-rs there is also the improvement in instructions from using the sized dealloc apis: https://github.com/rust-lang/rust/pull/83152#issuecomment-799755753, from before I added the tikv-jemalloc change, seems particularly good for rustdoc

**Comment by joshtriplett** [comment]
This seems like an excellent improvement in several ways; I'm particularly excited about the decreased memory usage. I think this would be reasonable to merge, as long as there's no objection to the use of the tikv crates.

File: src/librustdoc/lib.rs [hunk 2]
```diff
@@ -112,7 +122,45 @@ mod theme;
 mod visit_ast;
 mod visit_lib;
 
+// FIXME(guswynn): copy documentation from rustc/src/main.rs
+#[cfg(feature = "jemalloc")]
+#[global_allocator]
+static ALLOC: jemallocator::Jemalloc = jemallocator::Jemalloc;
```
Comment (jq-rs) [inline src/librustdoc/lib.rs:128]: This part is still missing tikv-update?

**Comment by jq-rs** [comment]
@guswynn I am still a bit puzzled which part of the implementation actually calls sized dealloc here, or was it so that rustdoc did not use `jemalloc` at all previously? In any case, I agree that rustdoc improves a lot with this change.

**Comment by mati865** [comment]
@jq-rs there are 2 things done in this PR, sized deallocation and enabling jemalloc for rustdoc.

Those 2 lines https://github.com/rust-lang/rust/pull/83152/files#diff-43914724af6e464c1da2171e4a9b6c7e607d5bc1203fa95c0ab85be4122605efR3-R4 lead to https://docs.rs/tikv-jemallocator/0.4.1/src/tikv_jemallocator/lib.rs.html#125.

Previously rustdoc did not use jemalloc at all.

**Comment by jq-rs** [comment]
@mati865 Makes sense now, `static ALLOC` setting maps the global allocator into use which eventually calls sized dealloc with `dealloc`. Thanks!

File: compiler/rustc/src/main.rs [hunk 3]
```diff
@@ -1,3 +1,11 @@
+// FIXME(guswynn): depending on perf results, document this thoroughly
```
Comment (Mark-Simulacrum) [inline compiler/rustc/src/main.rs:1]: Looks like this FIXME (and the related one in rustdoc) could use fleshing out.

**Review by Mark-Simulacrum — COMMENTED** [review]
I agree with @joshtriplett that I think we can move forward here; left one comment but once that's resolved r=me.

File: compiler/rustc/src/main.rs (hunk 3 above)
Comment (jyn514) [reply compiler/rustc/src/main.rs:1]: I would prefer rustdoc just links to the rustc docs, otherwise people will forgot to update them.

**Comment by guswynn** [comment]
@Mark-Simulacrum I think this should be ready now!

**Comment by Mark-Simulacrum** [comment]
@bors r+

Thanks so much!

**Comment by guswynn** [comment]
is that a spurious failure

**Comment by jyn514** [comment]
Looks like it died in the middle of a test, there's no logs:

image

@bors retry

**Comment by LifeIsStrange** [comment]
@guswynn hi, jemalloc is working on an unofficial release channel which would update every 1-3 months. 
They promise to do a less extensive but best effort performance and stability testing, for the x86-64 Linux platform. 
So for this platform it could make sense to use such pre-releases.

https://github.com/jemalloc/jemalloc/issues/2060

**Comment by nnethercote** [comment]
> Currently, on unix, rust uses jemalloc by importing its symbols to use them with the default, System (libc) global allocator. This PR switches its global alloc to `tikv-jemallocator`, which correctly uses sized deallocation (https://docs.rs/tikv-jemallocator/0.4.1/src/tikv_jemallocator/lib.rs.html#121-126). `tikv-jemallocator`, as far as I can tell, is a more up-to-date set of bindings to jemalloc than `jemallocator`

I know I'm late to the party, but: the first and third sentences here are correct, but the second sentence is mostly incorrect. Although this PR added a `#[global_allocator]`, it has no effect, and the mechanism described in the first sentence is still the one that causes jemalloc to be used.

> The perf results of this pr are in large part due to the version upgrade of jemalloc, but sized deallocation has a non-trivial improvement, particularly to rustdoc.

Sized deallocation doesn't come into the picture; the improvement was all because `tikv-jemalloc-sys` uses a newer version of jemalloc than `jemalloc-sys`. See #92222 for more details.

**Comment by guswynn** [comment]
@nnethercote One of the earlier perf runs up above in this pr used jemallocator (non-tikv) and more closely captured the much smaller, but still measurable effect of sized deallocation. The larger wins were the upgrade, as you mentioned; the pr description is likely outdated. Hearing that those sized dealloc apis are not being used, which led to #92222 Is very concerning to me, and I would need to dig deeper into why the `#[global_allocator]` attribute is not causing the sized apis to be used, if that is the case.

**Comment by nnethercote** [comment]
> @nnethercote One of the earlier perf runs up above in this pr used jemallocator (non-tikv) and more closely captured the much smaller, but still measurable effect of sized deallocation.

That perf run is [here](https://perf.rust-lang.org/compare.html?start=2ccf06302c08d7d4911aad40e66a9a3ee731c6f9&end=4f4b85262ea011b219ebd71d3da9ac82605df32e) and you'll see that almost every significant change is a `doc` build. (The two non-`doc` ones are likely noise.) You did add the use of jemalloc to rustdoc, but it's the `extern` stuff in `main` that has the effect of incorporating `jemalloc`, not the `#[global_allocator]`.

> I would need to dig deeper into why the `#[global_allocator]` attribute is not causing the sized apis to be used, if that is the case.

#92222's comment points to https://github.com/rust-lang/rust/pull/81782#issuecomment-784438001, which has an explanation.

**Comment by nnethercote** [comment]
BTW, this was still a great PR! Switching to tikv-jemallocator was a huge win, and adding it to rustdoc was also good. It's just that what actually happened and what everyone thought happened didn't quite match 😀

**Comment by guswynn** [comment]
@nnethercote oh WOW did not know about the complexity of how rustc is build

is decoupling `librustc_driver.so` and `libstd.so` something that anyone is looking at?

**Comment by nnethercote** [comment]
> is decoupling `librustc_driver.so` and `libstd.so` something that anyone is looking at?

Not that I know of. And given that [sized deallocation doesn't appear to be a win](https://github.com/rust-lang/rust/pull/92222#issuecomment-1000655042), there's not much motivation for it.
